### ISSUE TITLE
leashParticle

### ISSUE THREAD

**Issue by JoyStipe** [OpenCollar/opencollar#216]
Romka, I'm wondering why, at line 357-360, that change?
If I recall correctly, anything other than zero is evaluated as TRUE...
Basically, the setting number can be off, even a letter or word;
letters interpret as zero/FALSE, integers or even floats that aren't zero
should always interpret as TRUE.
That said, it doesn't matter if the entry is wrong type (shouldn't happen),
there will return a true or false statement = why I used one line without conditionals.

Just curious: If there is a valid reason for the change, ok. Otherwise it's less code & overhead/mem-use to just set the value as I did to begin with.

On another note: trying to find an ugly bug, default buttons in menu always return setting to old-school defaults, but tells you that the setting is what it should be in notify.

**Comment by JoyStipe** [comment]
Okay major bug #1: lines 913-924; 
insert this "SaveDefaultSettings(sToken, sValue);" at line 918
and remove 919-924.

the issue: doesn't save defaults from notecard/settings-list into it's local default list
effect: clicking "default" on any options or "resetdefaults" for all reverts them to old-default
rather than the value from settings defaults
reason: setting burst prefix is ALWAYS the name of the script it is for, there is no such thing as

```
default_<token>=<value>
```

*Addendum: "Global_" is used for specials that many scripts use, such as collar locked state.
Globals are the only exception to the scriptname_token=value rule.
Also, do not worry about 2 lists from settings - it only sends one, on resets, from the notecard

**Comment by JoyStipe** [comment]
found the other bug and that's a tricky one... old size setting was a float (which we just used for x and y), so the notecard saw it as a big ZERO_VECTOR & stored that

only ways to fix is change notecard entry (if exists) for size, or add a bunch of code to test for that one variable being a float and change it to vector when it is.
I vote for the former, and will work with wendy on replacing the server-side bmc with a changed notecard

**Comment by ghost** [comment]
line 357-360: it is necessary if in 'defaultsettings' will be written for Glow: 'off' instead '0' and 'on' instead '1' , in either case it is not necessary.

lines 919-924; it is necessary if the 'defaultsettings' has user defaults for leashParticle, like:
User_leashDefaults=Texture~name~Size~vector~Color~vector~Density~float~Gravity~float~Life~float~Glow~(on/off/1/0)

and YES, found BUG:  leashDefault in User_leashDefault  and leashDefaults in g_sDefaults = "leashDefaults_";

**Comment by ghost** [comment]
about particle size: 
I think it is better to use to change size on the X and Y, that makes particles rectangular instead of square.

**Comment by ghost** [comment]
p.s.  I'm from Russia, is obscure to write English 
stupid Google translator...

**Comment by JoyStipe** [comment]
weee :)   Okay, so - #357; good explanation, thank you. I like it thumbs up!

**Comment by JoyStipe** [comment]
#919 - this will take time to explain. Settings script has a specific format that I developed to prevent cross-contamination (2+ scripts accidentally getting assigned the same variable name for a very different purpose). Main reason for this= add-on/plug-in. Anyone can write and submit, not everyone digs thru every other script to verify each and every variable (look how many scripts we have; I wouldn't want to spend all that time just hunting for variable names to find a "free" one)
SO, how this works: every setting that is to be stored is put into this same exact format:

```
<ScriptName>_<Token>=<Value>
```

WHERE:
 ScriptName = the script this applies to, from after the first " - " to before the next (if exists) " - "
       (May have changed while I was out - I put in because I still wish to see the old ways of script
        version numbering hehehehe)
 Token = the variable identifier (not always the same as variable typecast name)
       This the script (ScriptName) will challenge/verify to compare & locate what variable is to be
         set.
 Value = Actual value to set 'Token' to.

Exceptions to rule: "Globals," such as lock status, are handled with

```
Global_<Token>=<Value>
```

Globals are special, in that they are such variables that are used by several scripts, and must remain constant across the collar subsystems (lock/unlock, etc)

```
Defaults_<Token>=<Value>
```

will never occur in/for settings script, unless there is an "OpenCollar - Defaults" script.

THAT said, the "bug" that you mention about User_leashDefaults (the list) is & was a commented line, i.e. not applicable to bugs, as it doesn't run. Please don't feel badly - been there, done that, burned the t-shirt (and parked the stuffed bunny by my pillow).

 the actual bug is in 913 - 924, as mentioned above ("Okay major bug #1: lines 913-924; "), and is fixed as I described (tested/verified in SL)

Further questions about this? Say the word & I'll park in SL sometime and we can chatter in real-time

Particle size setting change: Oh, it's fine I thought about it long ago & just stuck with what we had been using because I don't personally see desire for rectangular particle bits, and square is easier (lazy me) ... not that I can't foresee someone else _might_ ... So, yeah ok got the thumbs up from this old bat...

re: the Google Translator English/Russian - hehehhehe totally agree & I believe it butchered half of you profile. Anyways peeked at that yesterday, as well, so I already knew you're Russian. Well that or you just like playing with ascii symbols and making it look like Russian written language. Anyways, you do rather well with English so far - I'm betting much better than I ever did with Russian.

**Comment by JoyStipe** [comment]
Addendum to the settings description:
User Settings (in particular leash) are handled differently than defaults (from notecard), but are also handled, to an extent by settings, if user puts into notecard.
notecard entries contain more "tokens" than internal variables:
Designer settings - presets defined by a collar creator -

```
<Design>_<ScriptName>=<Token>~<Value>~<Token>~<Value>(etc)
```

 Design = third entry of description field in collar properties, that which follows the second tilde (~)
 ScriptName = script that these settings apply to
 Token = Variable identifier (name) that script uses
 Value = What to set the variable to

User settings - presets placed in notecard by the wearer -

```
User_<ScriptName>=<Token>~<Value>~<Token>~<Value>(etc)
```

settings script will compare user settings to design settings at initialization, and store appropriately, then "send" the set that is preferred (user setting of pref that can be entered/changed in notecard by user).
Settings made by the user via menu are not pushed to the settings script's "defaults" list, rather are stored in a temp list that the user can "see" in local chat via menu, and choose whether to add to notecard. Those settings are stored in settings, so that they will be restored on relog/tp/etc, but are kept separate from NC defaults in settings, so that those values can always be reverted back to.

... in way of language apologies - some words may not translate well, others not at all. Some are not proper English words, others are just typographical errors.

**Comment by ghost** [comment]
well, tell your way to save the 'leashParticle default settings' 
my head is stuck with ideas :(

**Comment by littlemousy** [comment]
How about leashparticle reading the defaultsettings notecard itself, Romka?  It's an ugly solution, but it is the only way it can tell if its a notecard setting or a changed setting it receives.

**Comment by ghost** [comment]
maybe use different tokens? 
like 'DefaultTexture' 'DefaultSize' e.t.c.

in defaultsettings card will be
User_leashParticle=DefaultTexture~deftexture~DefaultSize~defsize  ...

**Comment by JoyStipe** [comment]
the settings subsystem has been set to allow a ton of flexibility in the
NC. Please let me try again:
Designer settings in notecard:
collarid_leashParticle=TOKEN~VALUE~TOKEN~VALUE(etc)
where collarid_ is the part of collar description field after the second
tilde (~), for example, "jsbmc_" at the end of the BMC's description, is
that collar design id, the numbers in the middle are an old version set no
longer used, but handy for designer use. The first part let's not worry
about right now...
leashParticle is unique it is the NAME of the script that these settings
are for, without the "OpenCollar - "part. It can also work with older or
future OC versions where the version number of the script was added at end
(with space-tilde-space " ~ " between fields), as it checks for the part
after first " ~ ", and stops the field if another " ~ " occurs (as
originally written by me in evolution project).
TOKEN is each variable "name" that the using-script will find, and VALUE is
the setting for that token.
the format is that script name comes, followed by "=" (separator), TOKEN,
"~" (separator) [etc]
everything after that "=" is a variable + it's value, in a list.each
variable is a listed string, and each token is another. it is only the
order and separators that keep this method viable.
the very first part design_ ... this can be a collarid, Global_ (constant
value), or User_ (constant)
settings like jsbmc_Global=CType~implant ... these are special variables
used by many scripts, and always must be the same across them.
User_leashParticle=texture~chain .... these are user/wearer preferences.
they can be set to override design defaults, or just add to them as
defaults for that collar.

as example, I prefer magenta "rope" for texture, and thinner, denser (less
space in between "links"), and taughter ...
User_leashParticle=Texture~rope~Color~<1,0,1>~Gravity~0.04~Density~1.0~Size~<0.5,0.5,0>

I thought that I had explained this in-world better... please ask if this
still doesn't explain it.

On Tue, Apr 8, 2014 at 1:21 PM, Romka notifications@github.com wrote:

> maybe use different tokens?
> like 'DefaultTexture' 'DefaultSize' e.t.c.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/OpenCollar/OpenCollarUpdater/issues/216#issuecomment-39896895
> .

## 

~ A bug is a feature with seniority

**Comment by ghost** [comment]
okay,  with last BULD 41 3.960   let do that:
1. write to defaultsettings card own particle defaults User_leashParticle=.... (if it absent) 
2. change current leash particle settings via Customise menu (different from Default)
3. Leash to Grabby Post (or any post)
4. detach collar (or logout from SL) and attach collar back (or login to SL)
5. go to Leash Customise and ResetDefaults
6. Look at leash: it return to defaults sets in card or stay current?

**Comment by JoyStipe** [comment]
1 & 2 already exist in play & function currently, verified in SL.
3 verified works as intended.
4 ... and what is desired behavior?
5 current leash submenu has reset/default buttons already (and they work correctly), also verified recently.
6 returns to notecard defaults when a Default/Reset button is clicked in leash options submenus (current and intended behavior).

**Comment by JoyStipe** [comment]
/me scrapes the mold off the tiles...
